docs/manual/codespaces.md

```diff
@@ -0,0 +1,281 @@
+
+1. Click on the "Code" button in the GitHub UI.
+2. Start the Codespace on the `main` branch using the devcontainer settings defined in the `.devcontainer/devcontainer.json` file.
+3. Or open up an advanced menu
+4. And configure the Codespace settings manually.
+
+## Dev Container
+
+The key pillar behind Codespaces is the Containerlab' Dev Container image that defines the environment in which the lab will run. The Dev Container image is a Docker image that contains all the necessary tools and dependencies to run Containerlab and other tools that you might need in the lab.
+
+The following two files define Containerlab's Dev Container image:
+
+1. [devcontainer.json](https://github.com/srl-labs/containerlab/blob/main/.devcontainer/devcontainer.json) - the Dev Container configuration file that defines how the environment is built, configured and launched.
+2. [Dockerfile](https://github.com/srl-labs/containerlab/blob/main/.devcontainer/Dockerfile) - the Dockerfile that the Dev Container is built from.
+
+The resulting Dev Container image contains the following tools and dependencies:
+
+- containerlab binary installed via the deb repository
+- docker in docker setup
+- [gNMIc](https://gnmic.openconfig.net) and [gNOIc](https://gnoic.kmrd.dev) tools
+- Go SDK
+- Python 3 with pyenv
+- `gh` CLI tool
+- zsh shell with oh-my-zsh configuration
+- VS Code plugins
+
+## Catalog of Codespaces-enabled labs
+
+Having a codespace-enabled lab makes it super easy for users to start the lab and get to the fun part of the labbing.
+
+We encourage lab authors to add `codespaces` and `clab-topo` topics to the lab repository that supports Codespaces; that way users would be able to find the labs that they can run in Codespaces by following [**this link**](https://github.com/search?q=topic%3Aclab-topo+topic%3Acodespaces&type=repositories).
+
+## Tips, tricks and known issues
+
+### Authenticating with ghcr.io container registry
+
+If you happen to have a _private_ image that you want to use in Codespaces you can push this image to your personal GitHub registry.
+
+To be able to access a private image you would need to (re)authenticate with the `read:packages` token entitlement against the GitHub registry. Thankfully, it is a matter of a copy-paste exerceise.
+
+First, unset the existing token and request the one with `read:packages` capability:
+
+```bash
+unset GITHUB_TOKEN && gh auth refresh -s read:packages
+```
+
+You will be prompted to authenticate with your GitHub account and the new token will be generated for you. Then you can login to the registry using the newly acquired token:
+
+```bash
+gh auth token | \
+docker login ghcr.io -u $(cat /home/vscode/.config/gh/hosts.yml | \
+grep user: | awk '{print $2}') --password-stdin
+```
+
+<small>Of course you can also install tailscale or any other 0-tier VPN to access any other self-hosted private registry.</small>
```

Comment:
Replace lines 248 to 268 with:

If the image you are planning to use is not available on any public container registry, you can easily upload it to the Codespace when container starts using Codespace GUI or Github CLI. (Optional: we can add some instructions on how to upload images with GH CLI)  
After uploading the image, you can import it with `docker import ${IMAGE_FILENAME} ${IMAGE_NAME}`  
Alternatively you can use vendor provided tools to automate image download.

Everything written above makes total technical sense. However we still live in the world where most vendors resist publishing containers on any open (or private) registry and disrespect any disobedience.  
After all - let people poke vendors and add some pressure. This will not change if we use workarounds.

Reply:
this is not about the vendors, this is how to make use of the collection of private images one might have on their ghcr.io registry.

What you are proposing to add can be another subsection under the tips section.

Reply:
OK, got it know "private image" != "vendor image". I think is neutral enough. LGTM especially with vendor subsection for clarity.